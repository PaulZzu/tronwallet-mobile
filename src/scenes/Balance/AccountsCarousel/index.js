import React from 'react'
import Carousel from 'react-native-snap-carousel'
import { Dimensions, Alert, Clipboard } from 'react-native'
import ActionSheet from 'react-native-actionsheet'
import Toast from 'react-native-easy-toast'
import get from 'lodash/get'
import MixPanel from 'react-native-mixpanel'

import Card from './Card'

import { withContext } from '../../../store/context'
import { hideSecret } from '../../../utils/secretsUtils'
import { logSentry } from '../../../utils/sentryUtils'
import tl from '../../../utils/i18n'

const CURRENCIES = [
  tl.t('cancel'),
  'TRX',
  'USD',
  'EUR',
  'AUD',
  'GBP',
  'BTC',
  'ETH'
]

class AccountsCarousel extends React.Component {
  get price () {
    const { price, currency } = this.props.context
    return get(price, [currency, 'price'], 0)
  }

  _snapToNewAccount = () => {
    const { accounts } = this.props.context
    const createdItemPosition = accounts.length - 1
    // We set the state to load before the item is focused
    this._onSnapToItem(createdItemPosition)
    // Timeout needed for android
    setTimeout(() => this.carousel.snapToItem(createdItemPosition), 300)
  }

  _onSnapToItem = activeAccount => {
    const { setPublicKey, accounts } = this.props.context
    if (accounts.length) {
      const { address } = accounts[activeAccount]
      setPublicKey(address)
      MixPanel.trackWithProperties('Account Operation', { type: 'Switch account', address })
    }
  }

  // expose current index to parent
  currentIndex = () => this.carousel.currentIndex

  _alertHideAccount = address => {
    Alert.alert(tl.t('warning'), tl.t('hideAccount'), [
      { text: tl.t('cancel'), style: 'cancel' },
      { text: 'Remove', onPress: () => this._handleHideAccount(address) }
    ])
  }

  _handleHideAccount = async address => {
    try {
      const { pin, hideAccount } = this.props.context
      const nextAccountIndex = this.currentIndex() - 1
      await hideSecret(pin, address)
      this.carousel.snapToItem(nextAccountIndex)
      hideAccount(address)
      MixPanel.trackWithProperties('Account Operation', { type: 'Hide Account', address })
    } catch (error) {
      logSentry(error, 'Hide Account Handler')
      Alert.alert(tl.t('warning'), tl.t('error.hideAccount'))
    }
  }

  _handleCurrencyChange = async index => {
    if (index) {
      const currency = CURRENCIES[index]
      this.props.context.setCurrency(currency)
      MixPanel.trackWithProperties('Account Operation', { type: 'Set currency', currency })
    }
  }

  _onCopyAddress = async address => {
    try {
      await Clipboard.setString(address)
      this.refs.toast.show(tl.t('receive.clipboardCopied'))
      MixPanel.trackWithProperties('Account Operation', { type: 'Copy Address - Clipboard' })
    } catch (error) {
      logSentry(error, 'Copy Address - Clipboard')
    }
  }

  _formatAddress = address =>
    address
      .substring(0, 9)
      .concat('...')
      .concat(address.substring(25, address.length))

  _renderItem = ({ item, index }) => {
    if (!item) {
      return null
    }

    const { currency } = this.props.context

    const balance = item.balance || 0
    const tronPower = item.tronPower || 0
    const bandwidth = item.bandwidth || 0
    return (
      <Card
        name={item.name}
        address={item.address}
        price={this.price}
        balance={balance}
        tronPower={tronPower}
        bandwidth={bandwidth}
        currency={currency}
        showDeleteBtn={index !== 0}
        onCopy={this._onCopyAddress}
        onCurrencyPress={() => this.ActionSheet.show()}
        onDelete={this._alertHideAccount}
      />
    )
  }

  render () {
    const { accounts } = this.props.context
    return (
      !!accounts && (
        <React.Fragment>
          <Carousel
            ref={ref => {
              this.carousel = ref
            }}
            layout='default'
            enableMomentum
            decelerationRate={0.9}
            data={accounts}
            itemWidth={300}
            sliderWidth={Dimensions.get('window').width}
            slideStyle={{ paddingHorizontal: 6 }}
            renderItem={this._renderItem}
            onSnapToItem={this._onSnapToItem}
          />
          <ActionSheet
            ref={ref => {
              this.ActionSheet = ref
            }}
            title={tl.t('balance.chooseCurrency')}
            options={CURRENCIES}
            cancelButtonIndex={0}
            onPress={this._handleCurrencyChange}
          />
          <Toast
            ref='toast'
            position='center'
            fadeInDuration={750}
            fadeOutDuration={1000}
            opacity={0.8}
          />
        </React.Fragment>
      )
    )
  }
}

export default withContext(AccountsCarousel)
