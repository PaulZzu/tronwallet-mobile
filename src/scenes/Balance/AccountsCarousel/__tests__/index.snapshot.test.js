import React from 'react'
import 'react-native'
import { shallow } from 'enzyme'
import AccountsCarousel from './../index'

jest.mock('./../../../../utils/i18n')
jest.mock('./../../../../utils/secretsUtils')
jest.mock('./../../../../utils/sentryUtils')
jest.mock('./../../../../store/context')

describe('Account Carousel Snapshot', () => {
  let accountCarouselWrapper = null
  let context = null

  beforeEach(() => {
    context = {
      price: {
        TRX: {
          price: 1
        },
        TWX: {
          price: 2
        }
      },
      setCurrency: jest.fn(),
      currency: 'TRX',
      accounts: [{
        bandwidth: 1,
        tronPower: 50,
        balance: 150
      }]
    }

    const props = { context }

    accountCarouselWrapper = shallow(<AccountsCarousel {...props} />)
  })

  test('Should match snapshot', () => {
    expect(accountCarouselWrapper).toMatchSnapshot()
  })
})
