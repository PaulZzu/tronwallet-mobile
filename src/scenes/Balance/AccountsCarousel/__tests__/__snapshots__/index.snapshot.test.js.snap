// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Carousel Snapshot Should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AccountsCarousel
    context={
      Object {
        "accounts": Array [
          Object {
            "balance": 150,
            "bandwidth": 1,
            "tronPower": 50,
          },
        ],
        "currency": "TRX",
        "price": Object {
          "TRX": Object {
            "price": 1,
          },
          "TWX": Object {
            "price": 2,
          },
        },
        "setCurrency": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Carousel
          activeAnimationOptions={null}
          activeAnimationType="timing"
          activeSlideAlignment="center"
          activeSlideOffset={20}
          apparitionDelay={0}
          autoplay={false}
          autoplayDelay={5000}
          autoplayInterval={3000}
          callbackOffsetMargin={5}
          containerCustomStyle={Object {}}
          contentContainerCustomStyle={Object {}}
          data={
            Array [
              Object {
                "balance": 150,
                "bandwidth": 1,
                "tronPower": 50,
              },
            ]
          }
          decelerationRate={0.9}
          enableMomentum={true}
          enableSnap={true}
          firstItem={0}
          hasParallaxImages={false}
          inactiveSlideOpacity={0.7}
          inactiveSlideScale={0.9}
          inactiveSlideShift={0}
          itemWidth={300}
          layout="default"
          lockScrollTimeoutDuration={1000}
          lockScrollWhileSnapping={false}
          loop={false}
          loopClonesPerSide={3}
          onSnapToItem={[Function]}
          renderItem={[Function]}
          shouldOptimizeUpdates={true}
          slideStyle={
            Object {
              "paddingHorizontal": 6,
            }
          }
          sliderWidth={750}
          swipeThreshold={20}
          useScrollView={false}
          vertical={false}
        />,
        <ActionSheet
          cancelButtonIndex={0}
          onPress={[Function]}
          options={
            Array [
              "cancel",
              "TRX",
              "USD",
              "EUR",
              "AUD",
              "GBP",
              "BTC",
              "ETH",
            ]
          }
          title="balance.chooseCurrency"
        />,
        <Toast
          fadeInDuration={750}
          fadeOutDuration={1000}
          opacity={0.8}
          position="center"
          positionValue={120}
          textStyle={
            Object {
              "color": "white",
            }
          }
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "activeAnimationOptions": null,
          "activeAnimationType": "timing",
          "activeSlideAlignment": "center",
          "activeSlideOffset": 20,
          "apparitionDelay": 0,
          "autoplay": false,
          "autoplayDelay": 5000,
          "autoplayInterval": 3000,
          "callbackOffsetMargin": 5,
          "containerCustomStyle": Object {},
          "contentContainerCustomStyle": Object {},
          "data": Array [
            Object {
              "balance": 150,
              "bandwidth": 1,
              "tronPower": 50,
            },
          ],
          "decelerationRate": 0.9,
          "enableMomentum": true,
          "enableSnap": true,
          "firstItem": 0,
          "hasParallaxImages": false,
          "inactiveSlideOpacity": 0.7,
          "inactiveSlideScale": 0.9,
          "inactiveSlideShift": 0,
          "itemWidth": 300,
          "layout": "default",
          "lockScrollTimeoutDuration": 1000,
          "lockScrollWhileSnapping": false,
          "loop": false,
          "loopClonesPerSide": 3,
          "onSnapToItem": [Function],
          "renderItem": [Function],
          "shouldOptimizeUpdates": true,
          "slideStyle": Object {
            "paddingHorizontal": 6,
          },
          "sliderWidth": 750,
          "swipeThreshold": 20,
          "useScrollView": false,
          "vertical": false,
        },
        "ref": [Function],
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "cancelButtonIndex": 0,
          "onPress": [Function],
          "options": Array [
            "cancel",
            "TRX",
            "USD",
            "EUR",
            "AUD",
            "GBP",
            "BTC",
            "ETH",
          ],
          "title": "balance.chooseCurrency",
        },
        "ref": [Function],
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "fadeInDuration": 750,
          "fadeOutDuration": 1000,
          "opacity": 0.8,
          "position": "center",
          "positionValue": 120,
          "textStyle": Object {
            "color": "white",
          },
        },
        "ref": "toast",
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Carousel
            activeAnimationOptions={null}
            activeAnimationType="timing"
            activeSlideAlignment="center"
            activeSlideOffset={20}
            apparitionDelay={0}
            autoplay={false}
            autoplayDelay={5000}
            autoplayInterval={3000}
            callbackOffsetMargin={5}
            containerCustomStyle={Object {}}
            contentContainerCustomStyle={Object {}}
            data={
              Array [
                Object {
                  "balance": 150,
                  "bandwidth": 1,
                  "tronPower": 50,
                },
              ]
            }
            decelerationRate={0.9}
            enableMomentum={true}
            enableSnap={true}
            firstItem={0}
            hasParallaxImages={false}
            inactiveSlideOpacity={0.7}
            inactiveSlideScale={0.9}
            inactiveSlideShift={0}
            itemWidth={300}
            layout="default"
            lockScrollTimeoutDuration={1000}
            lockScrollWhileSnapping={false}
            loop={false}
            loopClonesPerSide={3}
            onSnapToItem={[Function]}
            renderItem={[Function]}
            shouldOptimizeUpdates={true}
            slideStyle={
              Object {
                "paddingHorizontal": 6,
              }
            }
            sliderWidth={750}
            swipeThreshold={20}
            useScrollView={false}
            vertical={false}
          />,
          <ActionSheet
            cancelButtonIndex={0}
            onPress={[Function]}
            options={
              Array [
                "cancel",
                "TRX",
                "USD",
                "EUR",
                "AUD",
                "GBP",
                "BTC",
                "ETH",
              ]
            }
            title="balance.chooseCurrency"
          />,
          <Toast
            fadeInDuration={750}
            fadeOutDuration={1000}
            opacity={0.8}
            position="center"
            positionValue={120}
            textStyle={
              Object {
                "color": "white",
              }
            }
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activeAnimationOptions": null,
            "activeAnimationType": "timing",
            "activeSlideAlignment": "center",
            "activeSlideOffset": 20,
            "apparitionDelay": 0,
            "autoplay": false,
            "autoplayDelay": 5000,
            "autoplayInterval": 3000,
            "callbackOffsetMargin": 5,
            "containerCustomStyle": Object {},
            "contentContainerCustomStyle": Object {},
            "data": Array [
              Object {
                "balance": 150,
                "bandwidth": 1,
                "tronPower": 50,
              },
            ],
            "decelerationRate": 0.9,
            "enableMomentum": true,
            "enableSnap": true,
            "firstItem": 0,
            "hasParallaxImages": false,
            "inactiveSlideOpacity": 0.7,
            "inactiveSlideScale": 0.9,
            "inactiveSlideShift": 0,
            "itemWidth": 300,
            "layout": "default",
            "lockScrollTimeoutDuration": 1000,
            "lockScrollWhileSnapping": false,
            "loop": false,
            "loopClonesPerSide": 3,
            "onSnapToItem": [Function],
            "renderItem": [Function],
            "shouldOptimizeUpdates": true,
            "slideStyle": Object {
              "paddingHorizontal": 6,
            },
            "sliderWidth": 750,
            "swipeThreshold": 20,
            "useScrollView": false,
            "vertical": false,
          },
          "ref": [Function],
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "cancelButtonIndex": 0,
            "onPress": [Function],
            "options": Array [
              "cancel",
              "TRX",
              "USD",
              "EUR",
              "AUD",
              "GBP",
              "BTC",
              "ETH",
            ],
            "title": "balance.chooseCurrency",
          },
          "ref": [Function],
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "fadeInDuration": 750,
            "fadeOutDuration": 1000,
            "opacity": 0.8,
            "position": "center",
            "positionValue": 120,
            "textStyle": Object {
              "color": "white",
            },
          },
          "ref": "toast",
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenThreeAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
